### PR TITLE
ci: nightly proving key generation workflow

### PR DESCRIPTION
**Pull request by baumstern**
This PR introduces a nightly proving key generation workflow. It will help us maintain the ability to generate the proving key and ensure that it remains functional even if no new changes are made to the circuits. It provides an additional layer of assurance for the codebase and allows others to build their own proving keys without any issues.

### CODE REVIEW

**Comment by samajammin** [comment]
@baumstern I assume this is aimed to solve https://github.com/privacy-scaling-explorations/maci/issues/803 ?

Moving forward, please provide PR descriptions & reference the relevant issue

**Comment by samajammin** [comment]
FYI @baumstern looks like this PR includes a bunch of miscellaneous commits. Why? Perhaps you didn't branch of the latest `dev`?

File: .github/workflows/nightly.yml [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          ref: master
+
+      - name: Use Node.js 16
+        uses: actions/setup-node@v3
+        with:
+          node-version: 16
```
Comment (samajammin) [inline .github/workflows/nightly.yml:21]: I believe we should use Node 18 now, correct? See https://github.com/privacy-scaling-explorations/maci/issues/775

Comment (baumstern) [reply .github/workflows/nightly.yml:21]: Yes for unit tests. But e2e test has problem with Node 18. See https://github.com/privacy-scaling-explorations/maci/pull/775#discussion_r1374742715

**Comment by baumstern** [comment]
> FYI @baumstern looks like this PR includes a bunch of miscellaneous commits. Why? Perhaps you didn't branch of the latest `dev`?

The reason for this is that the PR was created from the previous default branch, which was the `master`. [GitHub Actions schedule triggers only work on the default branch](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule). But the default branch has now been changed to the `dev` so things are sorted out well now.

File: .github/workflows/circuit-build.yml [hunk 2]
```diff
@@ -2,28 +2,26 @@ name: Circuit
 
 on:
   push:
-    branches: [ master, audit, dev ]
+    branches: [ master, dev ]
```
Comment (ctrlc03) [inline .github/workflows/circuit-build.yml:5]: good catch there

File: .github/workflows/nightly.yml [hunk 3]
```diff
@@ -0,0 +1,125 @@
+name: Nightly
+
+on:
+  schedule:
+    - cron: 0 0 * * *
+
+
+jobs:
+  generate-proving-key:
```
Comment (ctrlc03) [inline .github/workflows/nightly.yml:9]: small nit: would call it "generate-proving-keys" as it's > 1 key

File: .github/workflows/nightly.yml [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+      - name: Install dependencies
+        run: |
+          sudo apt-get update
+          sudo apt-get install --yes \
+            build-essential \
+            libgmp-dev \
+            libsodium-dev \
+            nasm \
+            nlohmann-json3-dev
```
Comment (ctrlc03) [inline .github/workflows/nightly.yml:30]: nlohmann-json3-dev is not needed anymore

File: .github/workflows/nightly.yml [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+      # To prevent `npm install` failure of circuit package, 
+      # it has to checkout manually because version of `circomlib` has pinned to a specific commit
+      - name: Checkout circomlib
```
Comment (ctrlc03) [inline .github/workflows/nightly.yml:34]: how come this is not something we do while working locally but it's needed here?

File: .github/workflows/nightly.yml [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+          sudo apt-get update
+          sudo apt-get install --yes \
+            build-essential \
+            libgmp-dev \
+            libsodium-dev \
+            nasm
+
+      # # To prevent `npm install` failure of circuit package, 
+      # # it has to checkout manually because version of `circomlib` has pinned to a specific commit
+      # - name: Checkout circomlib
```
Comment (samajammin) [inline .github/workflows/nightly.yml:34]: Why is this now commented out?

If not needed, can we delete this?

File: .github/workflows/nightly.yml (hunk 4 above)
Comment (baumstern) [reply .github/workflows/nightly.yml:31]: It appears that removing nlohmann-json3-dev leads to a compilation failure. Here's the error we encounter:

```
main.cpp:9:10: fatal error: nlohmann/json.hpp: No such file or directory
    9 | #include <nlohmann/json.hpp>
      |          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
```

File: .github/workflows/nightly.yml (hunk 5 above)
Comment (baumstern) [reply .github/workflows/nightly.yml]: he step to manually checkout circomlib was introduced in April 2022, as part of a pull request (see [PR #403](https://github.com/privacy-scaling-explorations/maci/pull/403/files)). The reason for this step was due to a specific requirement at the time. maci-circuits imports circomlib and pins it to a specific commit (as indicated [here in the package.json](https://github.com/privacy-scaling-explorations/maci/blob/dev/circuits/package.json#L25)). Back then, the direct pulling of this dependency with a pinned commit like https://github.com/weijiekoh/circomlib#ac85e82c1914d47789e2032fb11ceb2cfdd38a2b failed.

This issue could have been related to the behavior of npm.js or changes in GitHub's security policy, as outlined in this [GitHub blog post](https://github.blog/2021-09-01-improving-git-protocol-security-github/) or specific to the GitHub Actions Runner environment.

However, I've re-triggered the CI run without this manual checkout step and it worked smoothly, indicating that this step may no longer be necessary. It seems that the underlying issue has been resolved, either through updates to npm.js or GitHub's handling of dependencies.

File: .github/workflows/nightly.yml (hunk 4 above)
Comment (ctrlc03) [reply .github/workflows/nightly.yml:31]: mm I want to do some checks on a fresh VM actually. Looks like rapidnsark does not need this, but zkey-manager does, however where does it use it? is it here: https://github.com/privacy-scaling-explorations/zkey-manager/blob/master/ts/compile.ts#L102?

Comment (baumstern) [reply .github/workflows/nightly.yml:30]: It's needed when computing witness with c++: https://docs.circom.io/getting-started/computing-the-witness/

File: .github/workflows/nightly.yml (hunk 3 above)
Comment (baumstern) [reply .github/workflows/nightly.yml]: reflected in 20676ac36afb18c8d763b05e6d680aec52f0c518

File: .github/workflows/nightly.yml (hunk 6 above)
Comment (baumstern) [reply .github/workflows/nightly.yml]: Removed in 20676ac36afb18c8d763b05e6d680aec52f0c518

**Pull request merged**